TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/kafka/TrackingKafkaConsumer.java

```diff
@@ -0,0 +1,240 @@
+    private boolean hasPendingCommitsToSend;
+
+    public TrackingKafkaConsumer(Consumer<String, byte[]> kafkaConsumer, String topic,
+                                 Duration keepAliveInterval, Clock c) {
+        this.kafkaConsumer = kafkaConsumer;
+        this.topic = topic;
+        this.clock = c;
+        this.partitionToOffsetLifecycleTrackerMap = new HashMap<>();
+        this.nextSetOfCommitsMap = new HashMap<>();
+        this.lastTouchTimeRef = new AtomicReference<>(Instant.EPOCH);
+        this.keepAliveInterval = keepAliveInterval;
+    }
+
+    @Override
+    public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        safeCommit();
+        partitions.forEach(p->{
+            nextSetOfCommitsMap.remove(new TopicPartition(topic, p.partition()));
+            partitionToOffsetLifecycleTrackerMap.remove(p.partition());
+        });
+        if (hasPendingCommitsToSend) {
+            hasPendingCommitsToSend = partitionToOffsetLifecycleTrackerMap.values().stream()
+                    .anyMatch(olt -> !olt.isEmpty());
+        }
+        log.atWarn().setMessage(()->this+"partitions revoked for "+partitions.stream()
+                .map(p->p+"").collect(Collectors.joining(","))).log();
+    }
+
+    @Override public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        ++consumerConnectionGeneration;
+        partitions.forEach(p->partitionToOffsetLifecycleTrackerMap.computeIfAbsent(p.partition(),
+                x->new OffsetLifecycleTracker(consumerConnectionGeneration)));
+        log.atWarn().setMessage(()->this+"partitions added for "+partitions.stream()
+                .map(p->p+"").collect(Collectors.joining(","))).log();
+    }
+
+    public void close() {
+        log.atInfo().setMessage(()->"Kafka consumer closing.  " +
+                "Committing (implicitly by Kafka's consumer): " + nextCommitsToString()).log();
+        kafkaConsumer.close();
+    }
+
+    public Optional<Instant> getNextRequiredTouch() {
+        return hasPendingCommitsToSend ? Optional.of(lastTouchTimeRef.get().plus(keepAliveInterval)) : Optional.empty();
+    }
+
+    public void touch() {
+        log.trace("touch() called.");
+        pause();
+        try {
+            var records = kafkaConsumer.poll(Duration.ZERO);
+            if (!records.isEmpty()) {
+                throw new IllegalStateException("Expected no entries once the consumer was paused.  " +
+                        "This may have happened because a new assignment slipped into the consumer AFTER pause calls.");
+            }
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (RuntimeException e) {
+            log.atWarn().setCause(e).setMessage("Unable to poll the topic: " + topic + " with our Kafka consumer. " +
+                    "Swallowing and awaiting next metadata refresh to try again.").log();
+        } finally {
+            resume();
+        }
+        safeCommit();
```

Comment:
I wasn't expecting anything to be committed with touch(), what is happening here?

Reply:
We would like to see commits by synced to Kafka as soon as possible.  However, Kafka doesn't let us do that asynchronously, so we have to do it on the polling thread & we need to make sure that polls have recently been called.  Doing this in touch lets us do it more frequently (when commits are coming in) & let's us be sure that we should still be in the group because there was a recent poll().